src/lua/backtrace.h

```diff
@@ -0,0 +1,113 @@
+/*
+ * SPDX-License-Identifier: BSD-2-Clause
+ *
+ * Copyright 2010-2022, Tarantool AUTHORS, please see AUTHORS file.
+ */
+
+#pragma once
+
+#include "trivia/config.h"
+
+#ifdef ENABLE_BACKTRACE
+#include "core/backtrace.h"
+#include "core/say.h"
+
+#ifdef __cplusplus
+extern "C" {
+#endif /* __cplusplus */
+enum {
+	/* +2 for `backtrace_collect` and `backtrace_collect_current_stack` */
+	/* Maximal length of a Lua function  name. */
+	LUA_BACKTRACE_LUA_PROC_NAME_LEN_MAX = 64,
+	/* Maximal length of a Lua source file name. */
+	LUA_BACKTRACE_LUA_SRC_NAME_LEN_MAX = 64,
+	/* Maximal number of Lua frames collected. */
+	LUA_BACKTRACE_FRAME_COUNT_MAX = 64,
+};
+
+/*
+ * Lua frame information, derived from Core frame.
+ */
+struct lua_frame {
+	union {
+		/* Base frame type. */
+		struct core_frame core_frame;
+		/* Derived frame type. */
+		struct {
+			/* Lua function name. */
+			char proc_name[LUA_BACKTRACE_LUA_PROC_NAME_LEN_MAX];
+			/* Lua source file name. */
+			char src_name[LUA_BACKTRACE_LUA_SRC_NAME_LEN_MAX];
+			/* Line number inside Lua function. */
+			int line_no;
+		};
+	};
+	enum lua_backtrace_frame_type {
+		LUA_BACKTRACE_CORE_FRAME,
+		LUA_BACKTRACE_LUA_FRAME
+	} type;
+};
+
+/*
+ * Collection of Core and Lua frames.
+ */
+struct lua_backtrace {
+	/* Number of Lua frames collected. */
+	int frame_count;
+	/* Array of Lua frames. */
+	struct lua_frame frames[LUA_BACKTRACE_FRAME_COUNT_MAX];
+};
+
+struct fiber;
+struct lua_State;
+
+/*
+ * Callback for processing Lua frames.
+ *
+ * Returns status:
+ * 0 - continue processing.
+ * -1 - stop processing
+ */
+typedef int
+(*lua_frame_cb)(const struct lua_frame *, int frame_no, void *ctx);
+
+#ifndef __APPLE__
+/*
+ * Initialize Lua backtrace: setup Lua stack entry boundaries.
+ */
+void
+lua_backtrace_init(void);
+#endif /* __APPLE__ */
+
+/*
+ * Collect call stack of `fiber` (both C/C++ and Lua frames) to `bt`.
+ */
+void
+lua_backtrace_collect_frames(struct lua_backtrace *lua_bt, struct fiber *fiber,
```

Comment:
Using the `lua_` prefix might not be a good idea, because this prefix is used by the Lua library. This may result in name conflicts in future. May be better call `frame_lua`, `backtrace_lua`, etc.

Reply:
Fixed: changed all prefixes to `backtrace_lua_`.